### Reduction operation: the sum of the numbers in the range [0, a)

In [1]:
import time

def reduc_operation(a):
    """Compute the sum of the numbers in the range [0, a)."""
    x = 0
    for i in range(a):
        x += i
    return x

# Secuencial

value = 1000000

initialTime = time.time()
suma = reduc_operation(value)
finalTime = time.time()

print("Time taken by reduction operation:", (finalTime - initialTime), "seconds")

# Utilizando las operaciones mágicas de ipython
%timeit -r 2 reduc_operation(value)

print(f"\n \t Computing the sum of numbers in the range [0, value): {suma}\n")

Time taken by reduction operation: 0.031900882720947266 seconds
30.5 ms ± 50.3 μs per loop (mean ± std. dev. of 2 runs, 10 loops each)

 	 Computing the sum of numbers in the range [0, value): 499999500000



## a) Usando listas de Python

In [6]:
import time

# Crear lista con 10^6 elementos
value = 10**6
data_list = list(range(value))

# ==========================================
# Método 1: Bucle for
# ==========================================
# Definir función como buena práctica para el uso posterior de %timeit
def sum_with_for(data):
    total = 0
    for item in data:
        total += item
    return total

print("\n>>> Método 1: Bucle for con lista\n")

# Con librería time
start = time.time()
total = sum_with_for(data_list)
end = time.time()
print(f"  Tiempo (time): {end - start:.6f} segundos ({(end - start)*1000:.2f} ms)")

# Con %timeit
print("\n  Tiempo (%timeit):")
%timeit sum_with_for(data_list)

print(f"  Resultado: {total}")

# ==========================================
# Método 2: Función sum()
# ==========================================
print("\n>>> Método 2: Función sum() con lista\n")

# Con librería time
start = time.time()
total = sum(data_list)
end = time.time()
print(f"  Tiempo (time): {end - start:.6f} segundos ({(end - start)*1000:.2f} ms)")

# Con %timeit
print("\n  Tiempo (%timeit):")
%timeit sum(data_list)

print(f"  Resultado: {total}")


>>> Método 1: Bucle for con lista

  Tiempo (time): 0.024427 segundos (24.43 ms)

  Tiempo (%timeit):
23.4 ms ± 215 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)
  Resultado: 499999500000

>>> Método 2: Función sum() con lista

  Tiempo (time): 0.006186 segundos (6.19 ms)

  Tiempo (%timeit):
5.84 ms ± 21.2 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
  Resultado: 499999500000


## b) Usando arrays de NumPy

In [7]:
import time
import numpy as np

# Convertir la lista anterior a array de NumPy
data_array = np.array(data_list)

# ==========================================
# Método 1: Bucle for con array 
# ==========================================
print("\n>>> Método 1: Bucle for con array de NumPy\n")

# Con librería time
start = time.time()
total = sum_with_for(data_array)  # ← reutiliza la función del ejercicio (a)
end = time.time()
print(f"Tiempo (time):   {end - start:.6f} segundos ({(end - start)*1000:.2f} ms)")

# Con %timeit
print("\nTiempo (%timeit):")
%timeit sum_with_for(data_array)

print(f"\nResultado: {total}")

# ==========================================
# Método 2: numpy.sum()
# ==========================================
print("\n>>> Método 2: Función numpy.sum() con array\n")

# Con librería time
start = time.time()
total = np.sum(data_array)
end = time.time()
print(f"Tiempo (time):   {end - start:.6f} segundos ({(end - start)*1000:.2f} ms)")

# Con %timeit
print("\nTiempo (%timeit):")
%timeit np.sum(data_array)

print(f"\nResultado: {total}")


>>> Método 1: Bucle for con array de NumPy

Tiempo (time):   0.058228 segundos (58.23 ms)

Tiempo (%timeit):
57 ms ± 507 μs per loop (mean ± std. dev. of 7 runs, 10 loops each)

Resultado: 499999500000

>>> Método 2: Función numpy.sum() con array

Tiempo (time):   0.000401 segundos (0.40 ms)

Tiempo (%timeit):
243 μs ± 1.26 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Resultado: 499999500000


## c) Análisis comparativo de los resultados obtenidos por los tres procedimientos

### Comparación de tiempos (usando %timeit):

| Método | Estructura | Tiempo | Velocidad relativa |
|--------|-----------|--------|-------------------|
| **Código original** | range() interno | 30.5 ms | 1.0x (baseline) |
| **Bucle for** | Lista | 23.4 ms | 1.3x más rápido |
| **sum()** | Lista | 5.84 ms | **5.2x más rápido** |
| **Bucle for** | Array NumPy | 57 ms | **0.5x más lento** |
| **np.sum()** | Array NumPy | 243 µs | **125x más rápido** |


Los resultados obtenidos revelan diferencias significativas en el rendimiento entre las distintas aproximaciones evaluadas. El código original, que utiliza `range()` de forma interna, presenta un tiempo de ejecución de aproximadamente 30.5 ms, estableciendo la línea base para las comparaciones.

Al trabajar con listas de Python, se observa una mejora notable al emplear la función `sum()` integrada (5.84 ms) frente al bucle for manual (23.4 ms), logrando una aceleración de aproximadamente 5.2 veces respecto al código original. Este comportamiento se debe a que las funciones built-in de Python están optimizadas a nivel de C, minimizando el overhead del intérprete.

Sin embargo, los resultados más reveladores aparecen al introducir arrays de NumPy. Iterar sobre un array de NumPy mediante un bucle for explícito (57 ms) resulta **más lento** que hacerlo sobre una lista estándar. Este fenómeno se explica por el overhead de conversión de tipos entre Python y C en cada iteración, ya que NumPy debe convertir cada elemento del array (tipo NumPy) a un objeto Python antes de sumarlo.

La verdadera potencia de NumPy se manifiesta al utilizar sus funciones vectorizadas. La función `np.sum()` completa la operación en tan solo 243 microsegundos (0.243 ms), lo que representa una aceleración de **125 veces** respecto al código original. Esta mejora dramática se debe a que las operaciones vectorizadas están implementadas en C, aprovechan las instrucciones SIMD del procesador y evitan completamente el bucle interpretado en Python.

En conclusión, para cálculo científico con Python es fundamental utilizar las capacidades de vectorización que ofrece NumPy mediante sus funciones optimizadas, evitando en la medida de lo posible la iteración explícita sobre arrays.